force_wfmanager/utils/variable_names_registry.py

```diff
@@ -2,221 +2,326 @@
+    exec_layer_by_type = Dict(CUBAType, List(Variable))
+
+    #: Dictionary allowing the lookup of Variable objects with an associated
+    #: key. Practically, these keys are only used by update__variable_registry
+    #: to maintain the dictionary whenever its content is edited. Therefore
+    #: the dictionary itself remains private, whilst the Variables within the
+    #: dictionary are referred to externally, using the properties
+    #: available_variables, available_variable_names and
+    #: available_variables_by_type
+    _variable_registry = Dict(Unicode, Dict(Unicode, Variable))
 
     # -------------
     #   Properties
     # -------------
 
-    #: A list of type lookup dictionaries, with one dictionary for each
-    #: execution layer
-    available_variables_by_type = Property(
-        List(exec_layer_by_type),
-        depends_on="available_output_variables_stack,"
-                   "available_input_variables_stack"
-        )
+    #: A list containing the values of variable_registry dictionary
+    available_variables = Property(
+        List(Variable),
+        depends_on='_variable_registry'
+    )
+
+    #: A list containing a reference to each variable available_variables
+    variable_refs = Property(
+        List(Unicode),
+        depends_on='available_variables.[name,type]')
 
     #: Same structure as available_output_variables_stack, but this contains
-    #: the cumulated information. From the example above, this would contain::
-    #:
-    #:     [["Vol_A", "Vol_B"],
-    #:      ["Vol_A", "Vol_B"],
-    #:      ["Vol_A", "Vol_B", "Pressure_A"]]
-    #:
-    available_variables = Property(
+    #: the cumulated information.
+    available_variable_names = Property(
         List(List(Identifier)),
-        depends_on="available_output_variables_stack,"
-                   "available_input_variables_stack"
+        depends_on="available_variables"
     )
 
-    #: Gives only the names of the variables that are produced by data sources.
-    #: It does not include MCO parameters.
-    data_source_outputs = Property(
-        List(Identifier), depends_on="available_output_variables_stack"
-    )
-    data_source_inputs = Property(
-        List(Identifier), depends_on="available_input_variables_stack"
-    )
+    #: A list of type lookup dictionaries, with one dictionary for each
+    #: execution layer
+    available_variables_by_type = Property(
+        List(exec_layer_by_type),
+        depends_on="available_variables"
+        )
 
     def __init__(self, workflow, *args, **kwargs):
         super(VariableNamesRegistry, self).__init__(*args, **kwargs)
         self.workflow = workflow
 
+    def __variable_registry_default(self):
+        """Default value for _variable_registry dictionary: split into two inner
+        dictionaries, holding defined (created by a DataSource output slot
+        and undefined (requiring to be created by a MCOParameter) Variables"""
+        return {'defined': {},
```

Comment:
I don't know if these are the most descriptive names (`defined` and `undefined`), so feel free to suggest alternatives